libp2p-swarm/src/transport.rs

```diff
@@ -58,10 +58,16 @@ pub trait Transport {
 	/// A future which indicates that we are currently dialing to a peer.
 	type Dial: IntoFuture<Item = Self::RawConn, Error = IoError>;
 
-	/// Listen on the given multi-addr.
+	/// Listen on the given multiaddr. Returns a stream of incoming connections, plus a modified
+	/// version of the `Multiaddr`. This new `Multiaddr` is the one that that should be advertised
+	/// to other nodes, instead of the one passed as parameter.
 	///
 	/// Returns the address back if it isn't supported.
-	fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, (Self, Multiaddr)>
+	///
+	/// > **Note**: The reason why we need to change the `Multiaddr` on success is to handle
+	/// > 			situations such as turning `/ip4/127.0.0.1/tcp/0` into
+	/// > 			`/ip4/127.0.0.1/tcp/<actual port>`.
```

Comment:
I wish the multiaddr spec actually used the `*` symbol for wildcards instead of assuming things like "0 is a wildcard for TCP"

Reply:
I think it should be a numeric, 0 is a standard for auto-assignment for port.